File: test/src/nn/Flux2Transformer2DModel.py
from utils import TestCase
import torch
from diffusers.models.transformers.transformer_flux2 import Flux2Transformer2DModel

class TestNNFlux2Transformer2DModel(TestCase):
    def test_minimal(self):
        model = Flux2Transformer2DModel(
            patch_size=1,
            in_channels=8,
            out_channels=8,
            num_layers=1,
            num_single_layers=1,
            attention_head_dim=4,
            num_attention_heads=2,          # inner_dim = 8
            joint_attention_dim=16,
            timestep_guidance_channels=8,
            axes_dims_rope=(2, 2),    # sums to head_dim (=4)
            guidance_embeds=False,
        )

        hidden_states = torch.randn(1, 2, 8)
        encoder_hidden_states = torch.randn(1, 3, 16)

        timestep = torch.tensor([0.5])

        img_ids = torch.tensor([
            [
                [0, 0, 0, 0],
                [0, 0, 0, 1],
            ]
        ])

        txt_ids = torch.tensor([
            [
                [0, 0, 0, 2],
                [0, 0, 0, 3],
                [0, 0, 0, 4],
            ]
        ])

        expected = model(
            hidden_states=hidden_states,
            encoder_hidden_states=encoder_hidden_states,
            timestep=timestep,
            img_ids=img_ids,
            txt_ids=txt_ids,
        ).sample

        actual = self.cli(
            "Flux2Transformer2DModel",

            "--patch_size", "1",
            "--in_channels", "8",
            "--out_channels", "8",
            "--num_layers", "1",
            "--num_single_layers", "1",
            "--attention_head_dim", "4",
            "--num_attention_heads", "2",
            "--joint_attention_dim", "16",
            "--timestep_guidance_channels", "8",
            "--axes_dims_rope", "2",
            "--axes_dims_rope", "2",
            "--guidance_embeds", "false",

            "--hidden_states", str(hidden_states.tolist()),
            "--encoder_hidden_states", str(encoder_hidden_states.tolist()),
            "--timestep", str(timestep.tolist()),
            "--img_ids", str(img_ids.tolist()),
            "--txt_ids", str(txt_ids.tolist()),

            "--param-norm_out-linear-weight", str(model.norm_out.linear.weight.tolist()),
            "--param-proj_out-weight", str(model.proj_out.weight.tolist()),
            "--param-single_transformer_blocks-0-attn-norm_k-weight", str(model.single_transformer_blocks[0].attn.norm_k.weight.tolist()),
            "--param-single_transformer_blocks-0-attn-norm_q-weight", str(model.single_transformer_blocks[0].attn.norm_q.weight.tolist()),
            "--param-single_transformer_blocks-0-attn-to_qkv_mlp_proj-weight", str(model.single_transformer_blocks[0].attn.to_qkv_mlp_proj.weight.tolist()),
            "--param-single_transformer_blocks-0-attn-to_out-weight", str(model.single_transformer_blocks[0].attn.to_out.weight.tolist()),
            "--param-transformer_blocks-0-ff_context-linear_out-weight", str(model.transformer_blocks[0].ff_context.linear_out.weight.tolist()),
            "--param-transformer_blocks-0-ff_context-linear_in-weight", str(model.transformer_blocks[0].ff_context.linear_in.weight.tolist()),
            "--param-transformer_blocks-0-ff-linear_out-weight", str(model.transformer_blocks[0].ff.linear_out.weight.tolist()),
            "--param-transformer_blocks-0-ff-linear_in-weight", str(model.transformer_blocks[0].ff.linear_in.weight.tolist()),
            "--param-transformer_blocks-0-attn-to_add_out-weight", str(model.transformer_blocks[0].attn.to_add_out.weight.tolist()),
            "--param-transformer_blocks-0-attn-add_v_proj-weight", str(model.transformer_blocks[0].attn.add_v_proj.weight.tolist()),
            "--param-transformer_blocks-0-attn-add_k_proj-weight", str(model.transformer_blocks[0].attn.add_k_proj.weight.tolist()),
            "--param-transformer_blocks-0-attn-norm_added_k-weight", str(model.transformer_blocks[0].attn.norm_added_k.weight.tolist()),
            "--param-transformer_blocks-0-attn-norm_added_q-weight", str(model.transformer_blocks[0].attn.norm_added_q.weight.tolist()),
            "--param-transformer_blocks-0-attn-to_out-0-weight", str(model.transformer_blocks[0].attn.to_out[0].weight.tolist()),
            "--param-transformer_blocks-0-attn-norm_q-weight", str(model.transformer_blocks[0].attn.norm_q.weight.tolist()),
            "--param-transformer_blocks-0-attn-add_q_proj-weight", str(model.transformer_blocks[0].attn.add_q_proj.weight.tolist()),
            "--param-transformer_blocks-0-attn-to_v-weight", str(model.transformer_blocks[0].attn.to_v.weight.tolist()),
            "--param-transformer_blocks-0-attn-norm_k-weight", str(model.transformer_blocks[0].attn.norm_k.weight.tolist()),
            "--param-transformer_blocks-0-attn-to_k-weight", str(model.transformer_blocks[0].attn.to_k.weight.tolist()),
            "--param-transformer_blocks-0-attn-to_q-weight", str(model.transformer_blocks[0].attn.to_q.weight.tolist()),
            "--param-context_embedder-weight", str(model.context_embedder.weight.tolist()),
            "--param-single_stream_modulation-linear-weight", str(model.single_stream_modulation.linear.weight.tolist()),
            "--param-double_stream_modulation_txt-linear-weight", str(model.double_stream_modulation_txt.linear.weight.tolist()),
            "--param-double_stream_modulation_img-linear-weight", str(model.double_stream_modulation_img.linear.weight.tolist()),
            "--param-time_guidance_embed-timestep_embedder-linear_2-weight", str(model.time_guidance_embed.timestep_embedder.linear_2.weight.tolist()),
            "--param-time_guidance_embed-timestep_embedder-linear_1-weight", str(model.time_guidance_embed.timestep_embedder.linear_1.weight.tolist()),
            "--param-x_embedder-weight", str(model.x_embedder.weight.tolist()),
        )

        self.assertTensors(actual, [expected])

    def test_guidance(self):
        model = Flux2Transformer2DModel(
            patch_size=1,
            in_channels=8,
            out_channels=8,
            num_layers=1,
            num_single_layers=1,
            attention_head_dim=4,
            num_attention_heads=2,
            joint_attention_dim=16,
            timestep_guidance_channels=8,
            axes_dims_rope=(2, 2),
            guidance_embeds=True,
        )

        hidden_states = torch.randn(1, 2, 8)
        encoder_hidden_states = torch.randn(1, 3, 16)

        timestep = torch.tensor([0.5])
        guidance = torch.tensor([1.0])

        img_ids = torch.tensor([
            [
                [0, 0, 0, 0],
                [0, 0, 0, 1],
            ]
        ])

        txt_ids = torch.tensor([
            [
                [0, 0, 0, 2],
                [0, 0, 0, 3],
                [0, 0, 0, 4],
            ]
        ])

        expected = model(
            hidden_states=hidden_states,
            encoder_hidden_states=encoder_hidden_states,
            timestep=timestep,
            guidance=guidance,
            img_ids=img_ids,
            txt_ids=txt_ids,
        ).sample

        actual = self.cli(
            "Flux2Transformer2DModel",

            "--patch_size", "1",
            "--in_channels", "8",
            "--out_channels", "8",
            "--num_layers", "1",
            "--num_single_layers", "1",
            "--attention_head_dim", "4",
            "--num_attention_heads", "2",
            "--joint_attention_dim", "16",
            "--timestep_guidance_channels", "8",
            "--axes_dims_rope", "2",
            "--axes_dims_rope", "2",
            "--guidance_embeds", "true",

            "--hidden_states", str(hidden_states.tolist()),
            "--encoder_hidden_states", str(encoder_hidden_states.tolist()),
            "--timestep", str(timestep.tolist()),
            "--guidance", str(guidance.tolist()),
            "--img_ids", str(img_ids.tolist()),
            "--txt_ids", str(txt_ids.tolist()),

            "--param-norm_out-linear-weight", str(model.norm_out.linear.weight.tolist()),
            "--param-proj_out-weight", str(model.proj_out.weight.tolist()),
            "--param-single_transformer_blocks-0-attn-norm_k-weight", str(model.single_transformer_blocks[0].attn.norm_k.weight.tolist()),
            "--param-single_transformer_blocks-0-attn-norm_q-weight", str(model.single_transformer_blocks[0].attn.norm_q.weight.tolist()),
            "--param-single_transformer_blocks-0-attn-to_qkv_mlp_proj-weight", str(model.single_transformer_blocks[0].attn.to_qkv_mlp_proj.weight.tolist()),
            "--param-single_transformer_blocks-0-attn-to_out-weight", str(model.single_transformer_blocks[0].attn.to_out.weight.tolist()),
            "--param-transformer_blocks-0-ff_context-linear_out-weight", str(model.transformer_blocks[0].ff_context.linear_out.weight.tolist()),
            "--param-transformer_blocks-0-ff_context-linear_in-weight", str(model.transformer_blocks[0].ff_context.linear_in.weight.tolist()),
            "--param-transformer_blocks-0-ff-linear_out-weight", str(model.transformer_blocks[0].ff.linear_out.weight.tolist()),
            "--param-transformer_blocks-0-ff-linear_in-weight", str(model.transformer_blocks[0].ff.linear_in.weight.tolist()),
            "--param-transformer_blocks-0-attn-to_add_out-weight", str(model.transformer_blocks[0].attn.to_add_out.weight.tolist()),
            "--param-transformer_blocks-0-attn-add_v_proj-weight", str(model.transformer_blocks[0].attn.add_v_proj.weight.tolist()),
            "--param-transformer_blocks-0-attn-add_k_proj-weight", str(model.transformer_blocks[0].attn.add_k_proj.weight.tolist()),
            "--param-transformer_blocks-0-attn-norm_added_k-weight", str(model.transformer_blocks[0].attn.norm_added_k.weight.tolist()),
            "--param-transformer_blocks-0-attn-norm_added_q-weight", str(model.transformer_blocks[0].attn.norm_added_q.weight.tolist()),
            "--param-transformer_blocks-0-attn-to_out-0-weight", str(model.transformer_blocks[0].attn.to_out[0].weight.tolist()),
            "--param-transformer_blocks-0-attn-norm_q-weight", str(model.transformer_blocks[0].attn.norm_q.weight.tolist()),
            "--param-transformer_blocks-0-attn-add_q_proj-weight", str(model.transformer_blocks[0].attn.add_q_proj.weight.tolist()),
            "--param-transformer_blocks-0-attn-to_v-weight", str(model.transformer_blocks[0].attn.to_v.weight.tolist()),
            "--param-transformer_blocks-0-attn-norm_k-weight", str(model.transformer_blocks[0].attn.norm_k.weight.tolist()),
            "--param-transformer_blocks-0-attn-to_k-weight", str(model.transformer_blocks[0].attn.to_k.weight.tolist()),
            "--param-transformer_blocks-0-attn-to_q-weight", str(model.transformer_blocks[0].attn.to_q.weight.tolist()),
            "--param-context_embedder-weight", str(model.context_embedder.weight.tolist()),
            "--param-single_stream_modulation-linear-weight", str(model.single_stream_modulation.linear.weight.tolist()),
            "--param-double_stream_modulation_txt-linear-weight", str(model.double_stream_modulation_txt.linear.weight.tolist()),
            "--param-double_stream_modulation_img-linear-weight", str(model.double_stream_modulation_img.linear.weight.tolist()),
            "--param-time_guidance_embed-timestep_embedder-linear_1-weight", str(model.time_guidance_embed.timestep_embedder.linear_1.weight.tolist()),
            "--param-time_guidance_embed-timestep_embedder-linear_2-weight", str(model.time_guidance_embed.timestep_embedder.linear_2.weight.tolist()),
            "--param-time_guidance_embed-guidance_embedder-linear_1-weight", str(model.time_guidance_embed.guidance_embedder.linear_1.weight.tolist()),
            "--param-time_guidance_embed-guidance_embedder-linear_2-weight", str(model.time_guidance_embed.guidance_embedder.linear_2.weight.tolist()),
            "--param-x_embedder-weight", str(model.x_embedder.weight.tolist()),
        )

        self.assertTensors(actual, [expected])

    def test_multiple_blocks(self):
        model = Flux2Transformer2DModel(
            patch_size=1,
            in_channels=8,
            out_channels=8,
            num_layers=2,
            num_single_layers=2,
            attention_head_dim=4,
            num_attention_heads=2,          # inner_dim = 8
            joint_attention_dim=16,
            timestep_guidance_channels=8,
            axes_dims_rope=(2, 2),    # sums to head_dim (=4)
            guidance_embeds=False,
        )

        hidden_states = torch.randn(1, 2, 8)
        encoder_hidden_states = torch.randn(1, 3, 16)

        timestep = torch.tensor([0.5])

        img_ids = torch.tensor([
            [
                [0, 0, 0, 0],
                [0, 0, 0, 1],
            ]
        ])

        txt_ids = torch.tensor([
            [
                [0, 0, 0, 2],
                [0, 0, 0, 3],
                [0, 0, 0, 4],
            ]
        ])

        expected = model(
            hidden_states=hidden_states,
            encoder_hidden_states=encoder_hidden_states,
            timestep=timestep,
            img_ids=img_ids,
            txt_ids=txt_ids,
        ).sample

        actual = self.cli(
            "Flux2Transformer2DModel",

            "--patch_size", "1",
            "--in_channels", "8",
            "--out_channels", "8",
            "--num_layers", "2",
            "--num_single_layers", "2",
            "--attention_head_dim", "4",
            "--num_attention_heads", "2",
            "--joint_attention_dim", "16",
            "--timestep_guidance_channels", "8",
            "--axes_dims_rope", "2",
            "--axes_dims_rope", "2",
            "--guidance_embeds", "false",

            "--hidden_states", str(hidden_states.tolist()),
            "--encoder_hidden_states", str(encoder_hidden_states.tolist()),
            "--timestep", str(timestep.tolist()),
            "--img_ids", str(img_ids.tolist()),
            "--txt_ids", str(txt_ids.tolist()),

            "--param-norm_out-linear-weight", str(model.norm_out.linear.weight.tolist()),
            "--param-proj_out-weight", str(model.proj_out.weight.tolist()),
            "--param-single_transformer_blocks-0-attn-norm_k-weight", str(model.single_transformer_blocks[0].attn.norm_k.weight.tolist()),
            "--param-single_transformer_blocks-0-attn-norm_q-weight", str(model.single_transformer_blocks[0].attn.norm_q.weight.tolist()),
            "--param-single_transformer_blocks-0-attn-to_qkv_mlp_proj-weight", str(model.single_transformer_blocks[0].attn.to_qkv_mlp_proj.weight.tolist()),
            "--param-single_transformer_blocks-0-attn-to_out-weight", str(model.single_transformer_blocks[0].attn.to_out.weight.tolist()),
            "--param-transformer_blocks-0-ff_context-linear_out-weight", str(model.transformer_blocks[0].ff_context.linear_out.weight.tolist()),
            "--param-transformer_blocks-0-ff_context-linear_in-weight", str(model.transformer_blocks[0].ff_context.linear_in.weight.tolist()),
            "--param-transformer_blocks-0-ff-linear_out-weight", str(model.transformer_blocks[0].ff.linear_out.weight.tolist()),
            "--param-transformer_blocks-0-ff-linear_in-weight", str(model.transformer_blocks[0].ff.linear_in.weight.tolist()),
            "--param-transformer_blocks-0-attn-to_add_out-weight", str(model.transformer_blocks[0].attn.to_add_out.weight.tolist()),
            "--param-transformer_blocks-0-attn-add_v_proj-weight", str(model.transformer_blocks[0].attn.add_v_proj.weight.tolist()),
            "--param-transformer_blocks-0-attn-add_k_proj-weight", str(model.transformer_blocks[0].attn.add_k_proj.weight.tolist()),
            "--param-transformer_blocks-0-attn-norm_added_k-weight", str(model.transformer_blocks[0].attn.norm_added_k.weight.tolist()),
            "--param-transformer_blocks-0-attn-norm_added_q-weight", str(model.transformer_blocks[0].attn.norm_added_q.weight.tolist()),
            "--param-transformer_blocks-0-attn-to_out-0-weight", str(model.transformer_blocks[0].attn.to_out[0].weight.tolist()),
            "--param-transformer_blocks-0-attn-norm_q-weight", str(model.transformer_blocks[0].attn.norm_q.weight.tolist()),
            "--param-transformer_blocks-0-attn-add_q_proj-weight", str(model.transformer_blocks[0].attn.add_q_proj.weight.tolist()),
            "--param-transformer_blocks-0-attn-to_v-weight", str(model.transformer_blocks[0].attn.to_v.weight.tolist()),
            "--param-transformer_blocks-0-attn-norm_k-weight", str(model.transformer_blocks[0].attn.norm_k.weight.tolist()),
            "--param-transformer_blocks-0-attn-to_k-weight", str(model.transformer_blocks[0].attn.to_k.weight.tolist()),
            "--param-transformer_blocks-0-attn-to_q-weight", str(model.transformer_blocks[0].attn.to_q.weight.tolist()),
            "--param-single_transformer_blocks-1-attn-to_out-weight", str(model.single_transformer_blocks[1].attn.to_out.weight.tolist()),
            "--param-single_transformer_blocks-1-attn-norm_k-weight", str(model.single_transformer_blocks[1].attn.norm_k.weight.tolist()),
            "--param-single_transformer_blocks-1-attn-norm_q-weight", str(model.single_transformer_blocks[1].attn.norm_q.weight.tolist()),
            "--param-single_transformer_blocks-1-attn-to_qkv_mlp_proj-weight", str(model.single_transformer_blocks[1].attn.to_qkv_mlp_proj.weight.tolist()),
            "--param-transformer_blocks-1-ff_context-linear_out-weight", str(model.transformer_blocks[1].ff_context.linear_out.weight.tolist()),
            "--param-transformer_blocks-1-ff_context-linear_in-weight", str(model.transformer_blocks[1].ff_context.linear_in.weight.tolist()),
            "--param-transformer_blocks-1-ff-linear_out-weight", str(model.transformer_blocks[1].ff.linear_out.weight.tolist()),
            "--param-transformer_blocks-1-ff-linear_in-weight", str(model.transformer_blocks[1].ff.linear_in.weight.tolist()),
            "--param-transformer_blocks-1-attn-to_add_out-weight", str(model.transformer_blocks[1].attn.to_add_out.weight.tolist()),
            "--param-transformer_blocks-1-attn-add_v_proj-weight", str(model.transformer_blocks[1].attn.add_v_proj.weight.tolist()),
            "--param-transformer_blocks-1-attn-add_k_proj-weight", str(model.transformer_blocks[1].attn.add_k_proj.weight.tolist()),
            "--param-transformer_blocks-1-attn-norm_added_k-weight", str(model.transformer_blocks[1].attn.norm_added_k.weight.tolist()),
            "--param-transformer_blocks-1-attn-norm_added_q-weight", str(model.transformer_blocks[1].attn.norm_added_q.weight.tolist()),
            "--param-transformer_blocks-1-attn-to_out-0-weight", str(model.transformer_blocks[1].attn.to_out[0].weight.tolist()),
            "--param-transformer_blocks-1-attn-norm_q-weight", str(model.transformer_blocks[1].attn.norm_q.weight.tolist()),
            "--param-transformer_blocks-1-attn-add_q_proj-weight", str(model.transformer_blocks[1].attn.add_q_proj.weight.tolist()),
            "--param-transformer_blocks-1-attn-to_v-weight", str(model.transformer_blocks[1].attn.to_v.weight.tolist()),
            "--param-transformer_blocks-1-attn-norm_k-weight", str(model.transformer_blocks[1].attn.norm_k.weight.tolist()),
            "--param-transformer_blocks-1-attn-to_k-weight", str(model.transformer_blocks[1].attn.to_k.weight.tolist()),
            "--param-transformer_blocks-1-attn-to_q-weight", str(model.transformer_blocks[1].attn.to_q.weight.tolist()),
            "--param-context_embedder-weight", str(model.context_embedder.weight.tolist()),
            "--param-single_stream_modulation-linear-weight", str(model.single_stream_modulation.linear.weight.tolist()),
            "--param-double_stream_modulation_txt-linear-weight", str(model.double_stream_modulation_txt.linear.weight.tolist()),
            "--param-double_stream_modulation_img-linear-weight", str(model.double_stream_modulation_img.linear.weight.tolist()),
            "--param-time_guidance_embed-timestep_embedder-linear_2-weight", str(model.time_guidance_embed.timestep_embedder.linear_2.weight.tolist()),
            "--param-time_guidance_embed-timestep_embedder-linear_1-weight", str(model.time_guidance_embed.timestep_embedder.linear_1.weight.tolist()),
            "--param-x_embedder-weight", str(model.x_embedder.weight.tolist()),
        )

        self.assertTensors(actual, [expected])

File: src/diffusers/models/transformers/flux2/Flux2Transformer2DModel.cpp
#include "diffusers/models/transformers/flux2/Flux2Transformer2DModel.hpp"
#include "diffusers/models/transformers/flux2/Flux2PosEmbed.hpp"
#include "diffusers/models/transformers/flux2/Flux2TimestepGuidanceEmbeddings.hpp"
#include "diffusers/models/transformers/flux2/Flux2Modulation.hpp"
#include "diffusers/models/transformers/flux2/Flux2TransformerBlock.hpp"
#include "diffusers/models/transformers/flux2/Flux2SingleTransformerBlock.hpp"
#include "diffusers/models/normalization/AdaLayerNormContinuous.hpp"
#include "nn/Linear.hpp"
#include "nn/attention/ScaledDotProductAttention.hpp"
#include "nn/attention/FlashAttentionOp.hpp"
#include "ggml/GGUFLoaderVisitor.hpp"
#include <nlohmann/json.hpp>

using json = nlohmann::json;

Flux2Transformer2DModel::Flux2Transformer2DModel(const Flux2Transformer2DModel::Config& config) : Flux2Transformer2DModel(
    config.patch_size,
    config.in_channels,
    config.out_channels,
    config.num_layers,
    config.num_single_layers,
    config.attention_head_dim,
    config.num_attention_heads,
    config.joint_attention_dim,
    config.timestep_guidance_channels,
    config.mlp_ratio,
    config.axes_dims_rope,
    config.rope_theta,
    config.eps,
    config.guidance_embeds
) {

}

Flux2Transformer2DModel::Flux2Transformer2DModel(
    int64_t patch_size,
    int64_t in_channels,
    std::optional<int64_t> out_channels,
    int64_t num_layers,
    int64_t num_single_layers,
    int64_t attention_head_dim,
    int64_t num_attention_heads,
    int64_t joint_attention_dim,
    int64_t timestep_guidance_channels,
    float mlp_ratio,
    const std::vector<int64_t>& axes_dims_rope,
    int64_t rope_theta,
    float eps,
    bool guidance_embeds
) {
    auto inner_dim = num_attention_heads * attention_head_dim;

    // 1. Sinusoidal positional embedding for RoPE on image and text tokens
    modules["pos_embed"] = std::make_shared<Flux2PosEmbed>(rope_theta, axes_dims_rope);

    // 2. Combined timestep + guidance embedding
    modules["time_guidance_embed"] = std::make_shared<Flux2TimestepGuidanceEmbeddings>(
        timestep_guidance_channels,
        inner_dim,
        false,
        guidance_embeds
    );

    // 3. Modulation (double stream and single stream blocks share modulation parameters, resp.)
    // Two sets of shift/scale/gate modulation parameters for the double stream attn and FF sub-blocks
    modules["double_stream_modulation_img"] = std::make_shared<Flux2Modulation>(inner_dim, 2, false);
    modules["double_stream_modulation_txt"] = std::make_shared<Flux2Modulation>(inner_dim, 2, false);
    // Only one set of modulation parameters as the attn and FF sub-blocks are run in parallel for single stream
    modules["single_stream_modulation"] = std::make_shared<Flux2Modulation>(inner_dim, 1, false);

    // 4. Input projections
    modules["x_embedder"] = std::make_shared<Linear>(in_channels, inner_dim, false);
    modules["context_embedder"] = std::make_shared<Linear>(joint_attention_dim, inner_dim, false);

    // 5. Double Stream Transformer Blocks
    auto transformer_blocks = std::make_shared<ModuleList>(num_layers);
    modules["transformer_blocks"] = transformer_blocks;

    for (auto i = 0; i < transformer_blocks->size(); ++i)
        (*transformer_blocks)[i] =
            std::make_shared<Flux2TransformerBlock<ScaledDotProductAttention<FlashAttentionOp>>>(
                inner_dim,
                num_attention_heads,
                attention_head_dim,
                mlp_ratio,
                eps,
                false
            );


    // 6. Single Stream Transformer Blocks
    auto single_transformer_blocks = std::make_shared<ModuleList>(num_single_layers);
    modules["single_transformer_blocks"] = single_transformer_blocks;

    for (auto i = 0; i < single_transformer_blocks->size(); ++i)
        (*single_transformer_blocks)[i] =
            std::make_shared<Flux2SingleTransformerBlock<ScaledDotProductAttention<FlashAttentionOp>>>(
                inner_dim,
                num_attention_heads,
                attention_head_dim,
                mlp_ratio,
                eps,
                false
            );

    // 7. Output layers
    modules["norm_out"] = std::make_shared<AdaLayerNormContinuous<>>(inner_dim, inner_dim, false, eps, false);
    modules["proj_out"] = std::make_shared<Linear>(inner_dim, patch_size * patch_size * out_channels.value_or(in_channels), false);
}

Tensor Flux2Transformer2DModel::forward(
    Scope scope,
    Tensor hidden_states,
    Tensor encoder_hidden_states,
    Tensor timestep,
    Tensor img_ids,
    Tensor txt_ids,
    std::optional<Tensor> guidance,
    // TODO: support KV cache
    //Tensor joint_attention_kwargs: dict[str, Any] | None = None,
    //Tensor return_dict: bool = True,
    //Tensor kv_cache: "Flux2KVCache | None" = None,
    //Tensor kv_cache_mode: str | None = None,
    int64_t num_ref_tokens,
    float ref_fixed_timestep
) {
    auto num_txt_tokens = encoder_hidden_states.shape()[1];

    // 1. Calculate timestep embedding and modulation parameters
    timestep = (timestep * 1000.0f).to(hidden_states.dtype());

    if (guidance) {
        guidance = *guidance * 1000.0f;
        guidance = guidance->to(hidden_states.dtype());
    }

    auto time_guidance_embed = std::static_pointer_cast<Flux2TimestepGuidanceEmbeddings>(modules["time_guidance_embed"]);

    auto temb = time_guidance_embed->forward(scope, timestep, guidance);

    auto double_stream_modulation_img = std::static_pointer_cast<Flux2Modulation>(modules["double_stream_modulation_img"]);
    auto double_stream_modulation_txt = std::static_pointer_cast<Flux2Modulation>(modules["double_stream_modulation_txt"]);
    auto single_stream_modulation = std::static_pointer_cast<Flux2Modulation>(modules["single_stream_modulation"]);

    auto double_stream_mod_img = double_stream_modulation_img->forward(scope, temb);
    auto double_stream_mod_txt = double_stream_modulation_txt->forward(scope, temb);
    auto single_stream_mod = single_stream_modulation->forward(scope, temb);

    // TODO: KV extract mode: create cache and blend modulations for ref tokens

    // 2. Input projection for image (hidden_states) and conditioning text (encoder_hidden_states)
    auto x_embedder = std::static_pointer_cast<Linear>(modules["x_embedder"]);
    auto context_embedder = std::static_pointer_cast<Linear>(modules["context_embedder"]);

    hidden_states = x_embedder->forward(scope, hidden_states);
    encoder_hidden_states = context_embedder->forward(scope, encoder_hidden_states);

    // 3. Calculate RoPE embeddings from image and text tokens
    if (img_ids.ndim() == 3)
        img_ids = img_ids[0];
    if (txt_ids.ndim() == 3)
        txt_ids = txt_ids[0];

    // This is different from Python implementation because we'll compute RoPE directly using position IDs,
    // so embeddings setup is like this:
    auto concat_rotary_emb = std::make_pair(
        std::static_pointer_cast<Flux2PosEmbed>(modules["pos_embed"]),

        // [Nt, 4] + [Ni, 4] -> [Nt + Ni, 4]
        Tensor::cat({txt_ids, img_ids}, 0)
    );

    // TODO: 4. Build joint_attention_kwargs with KV cache info

    // 5. Double Stream Transformer Blocks
    auto transformer_blocks = std::static_pointer_cast<ModuleList>(modules["transformer_blocks"]);
    
    for (auto i = 0; i < transformer_blocks->size(); ++i) {
        auto block = std::static_pointer_cast<Flux2TransformerBlock<ScaledDotProductAttention<FlashAttentionOp>>>(
            (*transformer_blocks)[i]
        );

        auto [fwd_encoder_hidden_states, fwd_hidden_states] = block->forward(
            scope,
            hidden_states,
            encoder_hidden_states,
            double_stream_mod_img,
            double_stream_mod_txt,
            concat_rotary_emb
        );

        encoder_hidden_states = fwd_encoder_hidden_states;
        hidden_states = fwd_hidden_states;
    }

    // Concatenate text and image streams for single-block inference
    hidden_states = Tensor::cat({encoder_hidden_states, hidden_states}, 1);

    // TODO: Blend single block modulation for extract mode: [txt_mod, ref_mod, img_mod]
    // TOOD: Build single-block KV kwargs (single blocks need num_txt_tokens)

    // 6. Single Stream Transformer Blocks
    auto single_transformer_blocks = std::static_pointer_cast<ModuleList>(modules["single_transformer_blocks"]);

    for (auto i = 0; i < single_transformer_blocks->size(); ++i) {
        auto block = std::static_pointer_cast<Flux2SingleTransformerBlock<ScaledDotProductAttention<FlashAttentionOp>>>(
            (*single_transformer_blocks)[i]
        );

        auto [fwd_hidden_states, _] = block->forward(
            scope,
            hidden_states,
            std::nullopt,
            single_stream_mod,
            concat_rotary_emb
        );

        hidden_states = fwd_hidden_states;
    }

    // TODO: Remove text tokens (and ref tokens in extract mode) from concatenated stream

    hidden_states = hidden_states[{Tensor::Slice::all(), Tensor::Slice::range(num_txt_tokens, std::nullopt), Tensor::Slice::ellipsis()}];

    // 7. Output layers
    auto norm_out = std::static_pointer_cast<AdaLayerNormContinuous<>>(modules["norm_out"]);
    auto proj_out = std::static_pointer_cast<Linear>(modules["proj_out"]);

    hidden_states = norm_out->forward(scope, hidden_states, temb);
    auto output = proj_out->forward(scope, hidden_states);

    return output;
}


template <typename T>
void read(const json& j, const char* key, T& value) {
    if (!j.contains(key) || j[key].is_null())
        return;

    value = j[key].get<T>();
}

template <typename T>
void read(const json& j, const char* key, std::optional<T>& value) {
    if (!j.contains(key))
        return;

    if (j[key].is_null())
        value.reset();
    else
        value = j[key].get<T>();
}

Flux2Transformer2DModel::Config Flux2Transformer2DModel::Config::from_file(const std::filesystem::path& path) {
    std::ifstream file(path);

    if (!file.is_open())
        throw std::runtime_error("Failed to open configuration file: " + path.string());

    json j;
    file >> j;

    Flux2Transformer2DModel::Config cfg;

    read(j, "patch_size", cfg.patch_size);
    read(j, "in_channels", cfg.in_channels);
    read(j, "out_channels", cfg.out_channels);

    read(j, "num_layers", cfg.num_layers);
    read(j, "num_single_layers", cfg.num_single_layers);

    read(j, "attention_head_dim", cfg.attention_head_dim);
    read(j, "num_attention_heads", cfg.num_attention_heads);
    read(j, "joint_attention_dim", cfg.joint_attention_dim);
    read(j, "timestep_guidance_channels", cfg.timestep_guidance_channels);

    read(j, "mlp_ratio", cfg.mlp_ratio);
    read(j, "axes_dims_rope", cfg.axes_dims_rope);

    read(j, "rope_theta", cfg.rope_theta);
    read(j, "eps", cfg.eps);
    read(j, "guidance_embeds", cfg.guidance_embeds);

    return cfg;
}
